### ISSUE TITLE
Add kotlin Coroutines wrappers for operations 

### ISSUE THREAD

**Issue by digitalbuddha** [apollographql/apollo-android#1127]
Similar to rx support I'll be adding kotlin support to allow defferred apollo operations

**Comment by tylerbwong** [comment]
Hey @digitalbuddha, I'm curious about what this would look like for the most basic use-case. Would it be something like this?
```kt
suspend fun <T> ApolloCall<T>.await(): T? = suspendCancellableCoroutine { continuation ->
    enqueue(
        object : ApolloCall.Callback<T>() {
            override fun onResponse(response: Response<T>) = continuation.resume(response.data())
            override fun onCanceledError(ex: ApolloCanceledException) { continuation.cancel(ex) }
            override fun onFailure(e: ApolloException) = continuation.resumeWithException(e)
            override fun onNetworkError(e: ApolloNetworkException) = continuation.resumeWithException(e)
            override fun onParseError(e: ApolloParseException) = continuation.resumeWithException(e)
            override fun onHttpError(e: ApolloHttpException) = continuation.resumeWithException(e)
        }
    )
}
```

**Comment by martinbonnin** [comment]
I would expect more something along `Deferred<T>` ? 

Maybe 

     object CoroutineApollo {
         fun from( ApolloCall<T> call): Defferred<T>   
     } 

We actually need this as well, I might try to open a first PR in the coming days.

**Comment by martinbonnin** [comment]
Actually, since ApolloCall are more like `Observables` than `Singles`, I guess that translates in coroutines to `Channels` and not `Deferred` ?

**Comment by martinbonnin** [comment]
A very naive first draft: => https://github.com/martinbonnin/apollo-android/commit/92c3cce56bafab8f6eb6b1631417e88df13bc469

As said above, this doesn't work as soon as the call emits multiple values.

**Comment by martinbonnin** [comment]
First try at Channels:

    fun <T> ApolloCall<T>.toChannel(): Channel<Response<T>> {
        val channel = Channel<Response<T>>(Channel.UNLIMITED)
        enqueue(object : ApolloCall.Callback<T>() {
            override fun onResponse(response: Response<T>) {
                channel.offer(response)
            }
    
            override fun onFailure(e: ApolloException) {
                channel.close()
            }
        })
    
        return channel
    }
    
This seems to work pretty smoothly. I'm not 100% sure about error handling yet

**Comment by digitalbuddha** [comment]
Similar what to do with cancelations?

**Comment by martinbonnin** [comment]
Right, cancellation is another thing. 

I believe `ApolloCall.enqueue` cannot be canceled, is it ? That leaves us with `Channel.cancel()`. TBC if it has any guarantees in terms of `Channel.receive()` not working anymore (sorry if not the appropriate terminology, pretty new at coroutines here and even more at channels!)

**Comment by Tyler-Wong-CK** [comment]
I do believe calls can be canceled...the callback has `onCancelledError`. Would that work?

**Comment by martinbonnin** [comment]
`onCanceledError` defaults to `onFailure` so we could just forward the exception to `channel.close`. I just updated the code above to do this.

**Comment by martinbonnin** [comment]
New version that closes the channel when the call is completed:

    fun <T> ApolloCall<T>.toChannel(): Channel<Response<T>> {
        val channel = Channel<Response<T>>(Channel.UNLIMITED)
        enqueue(object : ApolloCall.Callback<T>() {
            override fun onResponse(response: Response<T>) {
                channel.offer(response)
            }

            override fun onFailure(e: ApolloException) {
                channel.close(e)
            }

            override fun onStatusEvent(event: ApolloCall.StatusEvent) {
                if (event == ApolloCall.StatusEvent.COMPLETED) {
                    channel.close()
                }
                
            }
        })

        return channel
    }

**Comment by digitalbuddha** [comment]
looks good to me, want to implement with a few tests being carried over from the rx impl?

**Comment by martinbonnin** [comment]
Right, I was looking for Rx tests in the rx modules but it looks like they are under `apollo-integration`. I'll look into that.

**Comment by martinbonnin** [comment]
Damn, `apollo-integration` is pure Java. Is it ok to pull Kotlin there ? Or should I write tests in `apollo-coroutines-support` ?

**Comment by digitalbuddha** [comment]
Kotlin in integration is fine with me. That module doesn't ship

**Comment by martinbonnin** [comment]
First drop there => https://github.com/apollographql/apollo-android/pull/1169/files 

I'm not too happy with it though. As @Tyler-Wong-CK mentioned above, calls/watchers can be canceled so we would need a way to forward the channel cancellation to the call/watcher cancel().

As of the pull request, the channel and all the interceptor chain continue running. That means watchers will continue watching forever....

**Comment by martinbonnin** [comment]
I asked the question on the kotlin slack and I was suggested to use `channel.invokeOnClose` which seems to do the job quite well. The pull request is considered working for me. Code reviews welcome :-)

**Comment by amitav13** [comment]
Eagerly awaiting this :)

**Comment by sav007** [comment]
Will be in next release.

**Comment by martinbonnin** [comment]
Mmm, good point. Thanks for noticing. @sav007 if it's not too late, I made a pull request there to fix the name: https://github.com/apollographql/apollo-android/pull/1199

I went with the plural form, I guess it's more used.

**Comment by timc-mediamonks** [comment]
What's the current status? I see PR 1199 has not been merged, but I cannot resolve either dependency atm

**Comment by martinbonnin** [comment]
@timbotnik  Sorry about that. The situation is a bit unstable right now. You can checkout the kotlin sample, which should work: https://github.com/martinbonnin/apollo-android/blob/3bb2e2fb07eb659c4189ebb4c66b857e6f7cd617/apollo-kotlin-sample/build.gradle#L42

What worked for me: 

    maven {
        url 'http://dl.bintray.com/apollographql/android'
    }

    dependencies {
        implementation "com.apollographql.apollo:apollo-coroutine-support:1.0.0-alpha5"
    }

**Comment by timc-mediamonks** [comment]
@martinbonnin ah, my mistake. I did not have the maven url added. Thanks

**Comment by martinbonnin** [comment]
We should certainly make this more clear in the README.md or move to mavenCentral altogether
